Enable embedding sources inside the sourcemaps json as sourcesContent

vegegoku opened the pull request:
This PR updates the rebased version of sourcemaps jar and other dependencies to support embedding the source files content into the source maps sourcesContent field.

requires : https://github.com/gwtproject/tools/pull/31

Fixes #9752 #9778 #9790

vegegoku committed:
Enable embedding sources inside the sourcemaps json as sourcesContent

niloc132 commented on dev/build.xml:
```diff
@@ -142,8 +143,9 @@
           <zipfileset
-              src="${gwt.tools.lib}/eclipse/jdtCompilerAdapter_3.17.0.v20190306-2240.jar"/>
-          <zipfileset src="${gwt.tools.lib}/guava/guava-19.0/guava-19.0-rebased.jar"/>
+              src="${gwt.tools.lib}/eclipse/jdtCompilerAdapter_3.32.0.v20221108-1853.jar"/>
```
if you're fixing this, should fix the line above too...

niloc132 commented on dev/core/src/com/google/gwt/core/linker/SymbolMapsLinker.java:
```diff
@@ -313,13 +326,46 @@ public ArtifactSet link(TreeLogger logger, LinkerContext context,
+                    }
                 });
+
+            StandardLinkerContext stdContext = (StandardLinkerContext) context;
```
As long as there is a cast here, might as well just access the module itself just once, rather than each time through the loop?

niloc132 reviewed (commented):
Some updates from private chat working on this feature:
 * We've learned that SDM won't be improved by this, as the various linking activity doesn't copy source files for in SDM, only in prod code with sourcemaps enabled. The "sourcesContent" part of this patch will only improve SDM in cases where the debugger loads many many sources, to the point where that network traffic could be time better spent reading a giant sourcemap file.
 * This is still a legit way to slightly improve prod build times for projects that deploy with sourcemaps (not many, but they exist), though it might be counterproductive for those only using sourcemaps for stack trace deobfuscation
 * For these reasons, we're leaning towards a config property in CompilerProperties.gwt.xml that would activate this new feature.
 * The patch still has value simply by updating the sourcemap library: both to potentially fix any issues in the last 8ish years since this was updated, and to drop the server-side runtime dependency on protobuf once and for all.

There is also other cleanup possible for compiler.useSourceMaps (since it should work in all browsers today, and not need extra wiring in CoreWithUserAgent), but let's leave that to another patch.

niloc132 commented:
Description should be updated to indicate that this fixes https://github.com/gwtproject/gwt/issues/9752.

vegegoku commented on dev/core/src/com/google/gwt/core/linker/SymbolMapsLinker.java:
```diff
@@ -59,6 +72,7 @@
 public class SymbolMapsLinker extends AbstractLinker {
 
   public static final String MAKE_SYMBOL_MAPS = "compiler.useSymbolMaps";
+  public static final String EMBED_SOURCE_MAPS_SOURCES = "compiler.embedSourceMapsSources";
```
This actually needs to be removed, this is now is read from the [ModuleDef](https://github.com/vegegoku/gwt/blob/b39f264bdf9fac87d93c034b9f83fd28d7c6efd8/dev/core/src/com/google/gwt/dev/cfg/ModuleDef.java#L637) which defaults to `false` if the property is not present.

niloc132 commented on dev/core/src/com/google/gwt/core/linker/SymbolMapsLinker.java:
```diff
@@ -335,6 +358,70 @@ public Object merge(String extKey, Object oldVal, Object newVal) {
+    section.parse(sourceMapString);
+    section.visitMappings(sourceMapGenerator::addMapping);
+
+    Iterator extensions = section.getExtensions().entrySet().iterator();
```
fix raw generics here, then the loop below won't need the cast?

niloc132 commented on dev/core/src/com/google/gwt/core/linker/SymbolMapsLinker.java:
```diff
@@ -335,6 +358,70 @@ public Object merge(String extKey, Object oldVal, Object newVal) {
+
+    ResourceLoader resourceLoader = ResourceLoaders.fromContextClassLoader();
+
+    Map<String, EmittedArtifact> generatedSources = new java.util.HashMap<>();
```
import HashMap - or probably more consistent with the rest of the codebase, use the factory method in Maps?

niloc132 commented on dev/core/src/com/google/gwt/core/linker/SymbolMapsLinker.java:
```diff
@@ -335,6 +358,70 @@ public Object merge(String extKey, Object oldVal, Object newVal) {
+
+    for (String sourceFileName : section.getOriginalSources()) {
+      String content;
+      try {
+        InputStream cis = loadSource(logger, sourceFileName, generatedSources,
+            resourceLoader);
```
Should be try-with-resources, right, so we are sure to close the file?

vegegoku replied:
The line after that is a check for a null and try to use a difference resource for loading the content, so I guess using the try with resource wont work here, but we can add a finally block to ensure we close the InputStream.

niloc132 commented on dev/core/src/com/google/gwt/core/linker/SymbolMapsLinker.java:
```diff
@@ -335,6 +358,70 @@ public Object merge(String extKey, Object oldVal, Object newVal) {
+      URL resource = resourceLoader.getResource(sourceFileName);
+      if (nonNull(resource)) {
+        URI uri = resource.toURI();
+        // Use Files.readAllBytes and Paths.get for Java 8 compatibility
```
Sorry, what is this comment saying?

vegegoku replied:
the line above should be actually removed as it is unused.

niloc132 commented on dev/core/src/com/google/gwt/core/ext/linker/impl/StandardLinkerContext.java:
```diff
@@ -537,6 +539,10 @@ public void produceOutput(TreeLogger logger, ArtifactSet artifacts,
     }
   }
 
+  public ModuleDef getModule() {
```
As we discussed when first working out how to plumb this, I don't really like exposing this here. I don't have a good enough answer to ask you to change it, so just documenting my thoughts in case something strikes one of us (or someone sees this PR in a few years and wants to know why on earth we made this change).

One option is to make a specific getter, `shouldEmbedSourcesInModuleSourcemaps()` sort of thing that just does `return module.isEmbedSourcesContent()`. That would better match getModuleName()/getModuleFunctionName()/getModuleLastModified() - though those three are part of LinkerContext itself. I don't think this should be added to LinkerContext. 

I dislike this option the least. However, it does suggest that we should do the same in CompilerContext - but CompilerContext has a bunch of other calls to getModule() already, so it feels a bit inconsistent.

Another option would be to remove the other three getters, and have all the callers instead call `getModule().getName()` and so on. This provides access to too much of ModuleDef I think, and it isn't immutable, so I really don't care for this. 

---

Separately, I don't love the name `isEmbedSourcesContent` as a method on ModuleDef, nor do I love that the method exists there at all. No other properties exist on ModuleDef itself, so it seems silly to have this one be special enough to be there. That said, it is needed by both the various pre/compile stages and linkers too, so it either needs to be somewhere very accessible, or needs to be duplicated - so I think I'm okay with it staying in ModuleDef _or_ pushing it down into Properties (so you have to call `getModule().getProperties.isBlahBlah()` - that does happen in other places, so it wouldn't be terrible). Renaming it: how about we get SourceMaps in the name somehow - shouldEmbedSourceMapContents() perhaps? sourceMapContentsEmbedEnabled()?

niloc132 reviewed (commented):
Looks like there may be more classpaths that _should_ be expanded, or code that needs to be corrected to stop walking classes it should ignore (annotations mostly?)

Example from the build:
```
requestfactory-server:
     [java] Could not find class file for com/google/auto/value/AutoValue
     [java] Could not find class file for com/google/auto/value/AutoValue$Builder
     [java] May 12, 2024 4:55:17 AM com.google.auto.value.AutoValue
     [java] SEVERE: Could not find class file for com/google/auto/value/AutoValue
     [java] May 12, 2024 4:55:17 AM com.google.auto.value.AutoValue.Builder
     [java] SEVERE: Could not find class file for com/google/auto/value/AutoValue$Builder
     [java] Could not find class file for com/google/errorprone/annotations/CanIgnoreReturnValue
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.CanIgnoreReturnValue
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/CanIgnoreReturnValue
     [java] Could not find class file for com/google/errorprone/annotations/DoNotCall
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.DoNotCall
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/DoNotCall
     [java] Could not find class file for com/google/errorprone/annotations/InlineMe
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.InlineMe
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/InlineMe
     [java] Could not find class file for com/google/errorprone/annotations/DoNotMock
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.DoNotMock
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/DoNotMock
     [java] Could not find class file for com/google/errorprone/annotations/InlineMeValidationDisabled
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.InlineMeValidationDisabled
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/InlineMeValidationDisabled
     [java] Could not find class file for com/google/errorprone/annotations/CompatibleWith
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.CompatibleWith
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/CompatibleWith
     [java] Could not find class file for com/google/errorprone/annotations/Immutable
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.Immutable
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/Immutable
     [java] Could not find class file for com/google/errorprone/annotations/ForOverride
     [java] Could not find class file for com/google/errorprone/annotations/CheckReturnValue
     [java] Could not find class file for com/google/errorprone/annotations/RestrictedApi
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.ForOverride
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/ForOverride
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.CheckReturnValue
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/CheckReturnValue
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.RestrictedApi
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/RestrictedApi
     [java] Could not find class file for com/google/errorprone/annotations/concurrent/GuardedBy
     [java] May 12, 2024 4:55:17 AM com.google.errorprone.annotations.concurrent.GuardedBy
     [java] SEVERE: Could not find class file for com/google/errorprone/annotations/concurrent/GuardedBy
```

vegegoku committed:
revert the removal of saveSources check and add the embed sources check

niloc132 reviewed (approved):
Approved, let's merge the tools PR, and remove the CI changes here, and give everyone else another day or so to see if they have any last minute feedback.

niloc132 commented:
Sorry, still lots of build warnings, I think more dependencies need to be added (or the RF jar extractor tool updated to otherwise handle this case): https://github.com/gwtproject/gwt/actions/runs/9196862375/job/25295956029?pr=9936#step:6:592

vegegoku committed:
workaround the warnings caused by upgrading source-mappings
error_prone will have the needed dependencies to include AutoValue and other annotations classes to the classpath and prevent the build warnings.

niloc132 commented:
Confirmed that this is fixed, waiting for full build to merge.

niloc132 commented:
Passing build: https://github.com/vegegoku/gwt/actions/runs/9975626940/job/27565849685

niloc132 commented on .github/workflows/full-check.yml:
```diff
@@ -29,7 +29,8 @@ jobs:
         uses: actions/checkout@v4
         with:
-          repository: 'gwtproject/tools'
+          repository: 'vegegoku/tools'
```
With the tools change merged, these need to be reverted to land this.

vegegoku committed:
revert CI changes after merging tools

niloc132 commented:
Full build after dropping `gwtproject/tools` workaround https://github.com/vegegoku/gwt/actions/runs/9978374423

jnehlmeier commented:
Running GWT including this change in production sometimes results in a NPE. See comment in https://github.com/gwtproject/gwt/issues/9931#issuecomment-2283421092
